### PR TITLE
Fix E2E code flow tests (fetching refs, updating version files)

### PR DESCRIPTION
**Pull request by premun**
Fixes problems introduced in #4181

- Fix how we look for git refs after we fetch from all remotes
- Fix getting the path to the `AllRepoVersions.props` file

https://github.com/dotnet/arcade-services/issues/3942

### CODE REVIEW

File: src/Microsoft.DotNet.Darc/DarcLib/LocalGitClient.cs [hunk 1]
```diff
@@ -22,6 +22,8 @@ namespace Microsoft.DotNet.DarcLib;
 /// </summary>
 public class LocalGitClient : ILocalGitClient
 {
+    private static readonly Regex ShaRegex = new(@"^[0-9a-fA-F]{40}$", RegexOptions.Compiled);
```
[the remark this answers is not in the document]
Comment (premun) [reply src/Microsoft.DotNet.Darc/DarcLib/LocalGitClient.cs]: Actually, I just realized it wouldn't work for local branches so I changed it a bit

File: src/Microsoft.DotNet.Darc/DarcLib/LocalGitClient.cs [hunk 2]
```diff
@@ -443,6 +442,22 @@ public async Task<string> BlameLineAsync(string repoPath, string relativeFilePat
         return result.StandardOutput.Trim().Split(' ').First();
     }
 
+    public async Task<bool> GitRefExists(string repoPath, string gitRef, CancellationToken cancellationToken = default)
```
Comment (dkurepa) [inline src/Microsoft.DotNet.Darc/DarcLib/LocalGitClient.cs:445]: is this needed? Aren't we making it so you can only flow builds, and we know the commit of a build

Comment (dkurepa) [reply src/Microsoft.DotNet.Darc/DarcLib/LocalGitClient.cs:445]: The bottom part, might never get used, if we're always passing commit shas

Comment (premun) [reply src/Microsoft.DotNet.Darc/DarcLib/LocalGitClient.cs:445]: The SHA detection was actually already there and working before this change. This change specifically makes it work with branches which we need to fetch (like the target PR branch). We need to fetch/verify if a branch exists in any remote

File: src/Microsoft.DotNet.Darc/DarcLib/VirtualMonoRepo/CloneManager.cs [hunk 3]
```diff
@@ -4,6 +4,7 @@
 using System;
 using System.Collections.Generic;
 using System.Linq;
+using System.Text.RegularExpressions;
```
Comment (dkurepa) [inline src/Microsoft.DotNet.Darc/DarcLib/VirtualMonoRepo/CloneManager.cs:7]: think we don't need this

**Review by dkurepa — APPROVED** [review]

**Pull request merged**